File: src/resources-packaged/xbl/orbeon/number/number.js
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
(function() {

    var $ = ORBEON.jQuery;
    var AS = ORBEON.xforms.server.AjaxServer;
    var Document = ORBEON.xforms.Document;

    YAHOO.namespace('xbl.fr');
    YAHOO.xbl.fr.Number = function() {};
    ORBEON.xforms.XBL.declareClass(YAHOO.xbl.fr.Number, 'xbl-fr-number');
    YAHOO.xbl.fr.Number.prototype = {

        xformsInputElement: null,
        xformsOutputElement: null,
        visibleInputElement: null,

        prefixElement: null,
        prefix: null,

        decimalSeparatorElement: null,
        decimalSeparator: null,

        groupingSeparatorElement: null,
        groupingSeparator: null,

        init: function() {

            this.xformsInputElement  = $(this.container).find('.xbl-fr-number-xforms-input')[0];
            this.xformsOutputElement = $(this.container).find('.xbl-fr-number-xforms-output')[0];
            this.visibleInputElement = $(this.container).find('.xbl-fr-number-visible-input')[0];

            // Properties
            // Find prefix based on class/control name, as this JS can be used with fr:number and fr:currency and properties use the control name
            var controlClassPrefix = null;
            var containerClasses = this.container.className.split(' ');
            for (var classIndex = 0; classIndex < containerClasses.length; classIndex++) {
                var currentClass = containerClasses[classIndex];
                if (currentClass.indexOf('xbl-fr-') == 0) {
                    controlClassPrefix = currentClass;
                    break;
                }
            }

            this.prefixElement            = $(this.container).find('.' + controlClassPrefix + '-prefix')[0];
            this.prefix                   = Document.getValue(this.prefixElement);
            this.decimalSeparatorElement  = $(this.container).find('.' + controlClassPrefix + '-decimal-separator')[0];
            this.decimalSeparator         = Document.getValue(this.decimalSeparatorElement);
            this.groupingSeparatorElement = $(this.container).find('.' + controlClassPrefix + '-grouping-separator')[0];
            this.groupingSeparator        = Document.getValue(this.groupingSeparatorElement);

            // Register listeners

            // Switch the input type after cleaning up the value for edition
            $(this.visibleInputElement).on('touchstart focus', _.bind(function(e) {

                this.visibleInputElement.value = this.numberToEditString(Document.getValue(this.xformsOutputElement));

                // With Firefox, changing the type synchronously interferes with the focus
                window.setTimeout(_.bind(function() {
                    $(this.visibleInputElement).attr('type', 'number');
                }, this), 0);
            }, this));

            // Restore input type, send the value to the server, and updates value after server response
            $(this.visibleInputElement).on('blur', _.bind(function(e) {

                // With Firefox, changing the type synchronously interferes with the focus
                window.setTimeout(_.bind(function() {
                    $(this.visibleInputElement).attr('type', 'text');
                }, this), 0);

                this.sendValueToServer();
                var formId = $(this.container).parents('form').attr('id');

                // Always update visible value with XForms value
                // - relying just value change event from server is not enough
                // - value change not dispatched if server value hasn't changed
                // - if visible changed, but XForms hasn't, we still need to show XForms value
                // - see: https://github.com/orbeon/orbeon-forms/issues/1026
                AS.nextAjaxResponse(formId).then(_.bind(this.updateWithServerValue, this));
            }, this));

            $(this.visibleInputElement).on('keypress', _.bind(function(e) {
                if (e.which == 13)
                    this.sendValueToServer();
            }, this));
        },

        setFocus: function() {
            this.visibleInputElement.focus();
         },

        sendValueToServer: function() {
            var newValue = this.visibleInputElement.value;
            Document.setValue(this.xformsInputElement, newValue);
        },

        numberToEditString: function(number) {
            var cleaned = number;

            // Remove spaces and grouping separators
            cleaned = cleaned.replace(new RegExp('[\\s' + this.groupingSeparator + ']', 'g'), '');

            // Remove prefix if present
            if (cleaned.indexOf(this.prefix) == 0)
                cleaned = cleaned.substring(this.prefix.length);

            var cleanedAsNumberString = cleaned.replace(new RegExp('[' + this.decimalSeparator + ']', 'g'), '.');

            return isNaN(Number(cleanedAsNumberString)) ? number : cleaned;
        },

        updateWithServerValue: function() {

            var numberFormattedValue = Document.getValue(this.xformsInputElement);
            var numberEditValue      = this.numberToEditString(numberFormattedValue);
            var hasFocus             = this.visibleInputElement == document.activeElement;

            this.visibleInputElement.value =
                hasFocus ?
                numberEditValue :
                numberFormattedValue;

            // Also update disabled because this might be called upon an iteration being moved, in which case all the control properties must be updated
            this.visibleInputElement.disabled = $(this.xformsInputElement).hasClass('xforms-readonly');
        },

        readonly: function() {
            this.visibleInputElement.disabled = true;
        },

        readwrite: function() {
            this.visibleInputElement.disabled = false;
        },

        parameterPrefixChanged: function() {
            this.prefix = Document.getValue(this.prefixElement);
            this.updateWithServerValue();
        },

        parameterDecimalSeparatorChanged: function() {
            this.decimalSeparator = Document.getValue(this.decimalSeparatorElement);
            this.updateWithServerValue();
        },

        parameterGroupingSeparatorChanged: function() {
            this.groupingSeparator = Document.getValue(this.groupingSeparatorElement);
            this.updateWithServerValue();
        }
    };
})();

File: form-runner-web/src/main/scala/org/orbeon/xbl/Number.scala
/**
  * Copyright (C) 2018 Orbeon, Inc.
  *
  * This program is free software; you can redistribute it and/or modify it under the terms of the
  * GNU Lesser General Public License as published by the Free Software Foundation; either version
  *  2.1 of the License, or (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
  * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  * See the GNU Lesser General Public License for more details.
  *
  * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
  */
package org.orbeon.xbl

import org.log4s.Logger
import org.orbeon.oxf.util.CoreUtils.*
import org.orbeon.oxf.util.LoggerFactory
import org.orbeon.web.DomEventNames
import org.orbeon.xforms.facade.XBL
import org.orbeon.xforms.{$, AjaxClient, AjaxEvent, Constants, EventListenerSupport}
import org.scalajs.dom
import org.scalajs.dom.html
import io.udash.wrappers.jquery.{JQueryEvent, JQueryPromise}
import org.scalajs.macrotaskexecutor.MacrotaskExecutor.Implicits.*

import scala.concurrent.duration.*
import scala.scalajs.js
import scala.scalajs.js.{Promise, UndefOr, timers, |}


object Number {

  private val logger: Logger = LoggerFactory.createLogger("org.orbeon.xbl.Number")
  private val TestNum = 1.1

  XBL.declareCompanion("fr|number",   js.constructorOf[NumberCompanion])
  XBL.declareCompanion("fr|currency", js.constructorOf[NumberCompanion])

  private class NumberCompanion(containerElem: html.Element) extends XBLCompanionWithState(containerElem) {

    companion =>

    import Private._
    import io.circe.generic.auto._
    import io.circe.{Decoder, Encoder}

    type State = NumberExternalValue

    val stateEncoder: Encoder[State] = implicitly[Encoder[State]]
    val stateDecoder: Decoder[State] = implicitly[Decoder[State]]
    private val eventListenerSupport = new EventListenerSupport {}

    var visibleInputElemOpt: Option[html.Input] = None

    override def init(): Unit = {

      logger.debug("init")

      val visibleInputElem = containerElem.querySelector(".xbl-fr-number-visible-input").asInstanceOf[html.Input]
      companion.visibleInputElemOpt = Some(visibleInputElem)

      // Switch the input type after cleaning up the value for edition
      eventListenerSupport.addListeners(visibleInputElem, List(DomEventNames.FocusIn, DomEventNames.TouchStart), (e: dom.Event) => {
        if (! isMarkedReadonly) {

          logger.debug(s"reacting to event ${e.`type`}")

          stateOpt foreach { state =>
            // Don't set value if not needed, so not to unnecessarily disturb the cursor position
            if (readValue(visibleInputElem, state.decimalSeparator) != state.editValue)
              writeValue(visibleInputElem, state.decimalSeparator, state.editValue)
          }

          setInputTypeIfNeeded("number")
        }
      })

      // Restore input type, send the value to the server, and updates value after server response
      eventListenerSupport.addListener(visibleInputElem, DomEventNames.FocusOut, (e: dom.Event) => {
        if (! isMarkedReadonly) {
          logger.debug(s"reacting to event ${e.`type`}")

          updateStateAndSendValueToServer(readValue)
          setInputTypeIfNeeded("text")

          // Always update visible value with XForms value:
          //
          // - relying just on value change event from server is not enough
          // - value change is not dispatched if the server value hasn't changed
          // - if the visible changed, but XForms hasn't, we still need to show XForms value
          // - see: https://github.com/orbeon/orbeon-forms/issues/1026
          //
          // So either:
          //
          // - the Ajax response called `updateWithServerValues()` and then `updateVisibleValue()`
          // - or it didn't, and we force `updateVisibleValue()` after the response is processed
          //
          // We also call `updateVisibleValue()` immediately in `updateStateAndSendValueToServer()` if the
          // edit value hasn't changed.
          //
          AjaxClient.allEventsProcessedF("number") foreach { _ =>
            updateVisibleValue()
          }
        }
      })

      eventListenerSupport.addListener(visibleInputElem, DomEventNames.KeyPress, (e: dom.KeyboardEvent) => {
        if (! isMarkedReadonly) {

          logger.debug(s"reacting to event ${e.`type`}")

          if (Set(10, 13)(e.keyCode)) {
            e.preventDefault()
            updateStateAndSendValueToServer(readValue)
            AjaxClient.fireEvent(
              AjaxEvent(
                eventName = DomEventNames.DOMActivate,
                targetId  = containerElem.id
              )
            )
          }
        }
      })
    }

    override def destroy(): Unit = {
      logger.debug("destroy")
      eventListenerSupport.clearAllListeners()
      companion.visibleInputElemOpt = None
    }

    override def xformsUpdateReadonly(readonly: Boolean): Unit = {
      logger.debug(s"xformsUpdateReadonly: $readonly")
      updateReadonly(readonly)
    }

    def xformsUpdateState(previousStateOpt: Option[State], newState: State): Unit = {

      val NumberExternalValue(displayValue, editValue, decimalSeparator) = newState

      logger.debug(s"updateWithServerValues: `$displayValue`, `$editValue`, `$decimalSeparator`")

      updateVisibleValue()

      // Also update disabled because this might be called upon an iteration being moved, in which
      // case all the control properties must be updated.
      updateReadonly(companion.isMarkedReadonly)
    }

    override def xformsFocus(): Unit = {
      logger.debug(s"xformsFocus")
      companion.visibleInputElemOpt foreach (_.focus())
    }

    override def setUserValue(newValue: String): UndefOr[Promise[Unit] | JQueryPromise[js.Function1[js.Any, js.Any], js.Any]] =
      updateStateAndSendValueToServer((_, _) => newValue)

    private object Private {

      private val hasToLocaleString =
        ! js.isUndefined(TestNum.asInstanceOf[js.Dynamic].toLocaleString)

      // On mobile, we set the field to `type="number"`, so the format of `value` always uses a `.`
      // as decimal separator
      private def mobileConversionNeeded(input: html.Input, decimalSeparator: Char): Boolean =
        input.`type` == "number" && decimalSeparator != '.'

      def updateReadonly(readonly: Boolean): Unit =
        visibleInputElemOpt foreach { visibleInputElem =>
          if (readonly)
            visibleInputElem.setAttribute("readonly", "readonly")
          else
            visibleInputElem.removeAttribute("readonly")
        }

      def updateStateAndSendValueToServer(read: (html.Input, Char) => String): Unit =
        visibleInputElemOpt foreach { visibleInputElem =>
          stateOpt foreach { state =>

            val visibleInputElemValue = read(visibleInputElem, state.decimalSeparator)

            val stateUpdated =
              updateStateAndSendValueToServerIfNeeded(
                newState       = state.copy(displayValue = visibleInputElemValue, editValue = visibleInputElemValue),
                valueFromState = _.editValue
              )

            if (! stateUpdated)
              updateVisibleValue()
          }
        }

      def updateVisibleValue(): Unit =
        companion.visibleInputElemOpt foreach { visibleInputElem =>

          def hasFocus = visibleInputElem eq dom.document.activeElement

          stateOpt foreach { state =>
            val newValue =
              if (! companion.isMarkedReadonly && hasFocus)
                state.editValue
              else
                state.displayValue
            writeValue(visibleInputElem, state.decimalSeparator, newValue)
          }
        }

      def setInputTypeIfNeeded(typeValue: String): Unit = {

        // Only switch to `type="number"` if the OS decimal separator is the same as the field decimal separator
        // See https://github.com/orbeon/orbeon-forms/issues/2545
        def hasNativeDecimalSeparator(separator: Char): Boolean =
          hasToLocaleString &&
            TestNum.asInstanceOf[js.Dynamic].toLocaleString().substring(1, 2).asInstanceOf[String] == separator.toString

        val changeType =
          dom.document.body.classList.contains(Constants.XFormsMobileClass) &&
            companion.stateOpt.exists(state => hasNativeDecimalSeparator(state.decimalSeparator))

        if (changeType) {
          // With Firefox, changing the type synchronously interferes with the focus
          timers.setTimeout(0.millis) {
            visibleInputElemOpt foreach { visibleInputElem =>
              $(visibleInputElem).attr("type", typeValue)
              // Set again the `input.value`, as otherwise we might loose the value (switching the type to `number`
              // if the current value is `1,2`), or be incorrect (switching the type to `text` if the current value
              // is `1.2` and the decimal separator `,`)
              updateVisibleValue()
            }
          }
        }
      }

      def readValue(input: html.Input, decimalSeparator: Char): String = {
        val convertFromMobile = mobileConversionNeeded(input, decimalSeparator)
        input.value.pipeIf(convertFromMobile, _.replace('.', decimalSeparator))
      }

      def writeValue(input: html.Input, decimalSeparator: Char, value: String): Unit = {
        val convertForMobile = mobileConversionNeeded(input, decimalSeparator)
        input.value = value.pipeIf(convertForMobile, _.replace(decimalSeparator, '.'))
      }
    }
  }
}
